### PR TITLE
Refactoring/services

### PR DESCRIPTION
**Pull request by jpontois**
**Route**
* Group
    * GET /groups/:user_id --> GET /groups/user/:user_id
* User
    * GET /user --> GET /users
    * PUT /user/user_id --> PUT /user
    * DELETE /user/user_id --> DELETE /user
    * Creation : GET /logout

**Controller**
* Made use of try catch structure to handle normal usage errors and systems errors.
* Creation of a custom error class to distinguish these two error types.
* Creation of services to allow easier interaction between different models.
* Creation of id validation before using them on a mongoose request.

**Model**
* Project
    * Admin field is now an user object with these fields
        * user_id
        * lastname
        * name
        * email

### CODE REVIEW

**Comment by jpontois** [comment]
**Additionnal feature and fixes after review**
* Fix group services that check the admin of a group
* Fix 2 function in Group controller that weren't working properly with wrong ID
    * CreateGroup
* Fix 2 function in project controller that weren't working properly with wrong ID
    * DeleteProject
    * UpdateProject

File: src/controllers/projectController.js [hunk 1]
```diff
@@ -1,98 +1,74 @@
 const mongoose = require('mongoose');
```
Comment (BuyukCem) [inline src/controllers/projectController.js:1]: can you delete unused mongoose import

File: src/controllers/projectController.js [hunk 2]
```diff
@@ -1,98 +1,74 @@
...
 const ProjectModel = require('../models/projectModel');
-const projectmodel = mongoose.model("Project");
-const groupController = require('../controllers/groupController')
-
+const groupServices = require('../services/groups.services')
+const projectServices = require('../services/project.services')
+const userServices = require('../services/users.services')
+const ApplicationError = require("../errors/application.errors");
+const validationParams = require('../utils/validationParams')
```
Comment (BuyukCem) [inline src/controllers/projectController.js:7]: can you delete unused validation Params

File: src/controllers/timerController.js [hunk 3]
```diff
@@ -3,95 +3,102 @@ const TimerModel = require('../models/timerModel');
...
-                        });
-                    }
-                });
+                    });
+                }
             }
-        }
-    });
+        });
+    } catch (e) {
```
Comment (BuyukCem) [inline src/controllers/timerController.js:53]: change name variable to error

File: src/controllers/projectController.js (hunk 1 above)
Comment (BuyukCem) [inline src/controllers/projectController.js:1]: can you delete unused mongoose import

File: src/controllers/projectController.js (hunk 2 above)
Comment (BuyukCem) [inline src/controllers/projectController.js:7]: can you delete unused variable: validationParams

File: src/controllers/timerController.js [hunk 4]
```diff
@@ -3,95 +3,102 @@ const TimerModel = require('../models/timerModel');
...
+    try {
+        TimerModel.find({project_id: req.params.id},(error, list) => {
+            if (error) {
+                throw new Error(error)
+            } else {
+                res.status(200);
+                res.json(list);
+            }
+        })
+    } catch (e) {
```
Comment (BuyukCem) [inline src/controllers/timerController.js:69]: change name variable to error

File: src/controllers/timerController.js [hunk 5]
```diff
@@ -3,95 +3,102 @@ const TimerModel = require('../models/timerModel');
...
+    try {
+        TimerModel.find({user_id: req.params.id}, (error, list) => {
+            if (error) {
+                throw new Error(error)
+            } else {
+                res.status(200);
+                res.json(list);
+            }
+        })
+    } catch (e) {
```
Comment (BuyukCem) [inline src/controllers/timerController.js:84]: change name variable to error

File: src/controllers/timerController.js [hunk 6]
```diff
@@ -3,95 +3,102 @@ const TimerModel = require('../models/timerModel');
...
+    try {
+        TimerModel.remove({"_id": req.params.id}, (error) => {
+            if (error) {
+                throw new Error(error)
+            } else {
+                res.status(200);
+                res.json({"message": "timer successfully removed"});
+            }
+        })
+    } catch (e) {
```
Comment (BuyukCem) [inline src/controllers/timerController.js:100]: change name variable to error

File: src/controllers/userController.js [hunk 7]
```diff
@@ -1,145 +1,183 @@
 const mongoose = require('mongoose');
-const UserModel = require('../models/userModel');
-const usermodel = mongoose.model("User");
-const jwt = require('../utils/jwt');
-const bcrypt = require('bcrypt');
-const {registerValidation, loginValidation, updateUserValidation} = require('../utils/validation.js');
-
-const groupController = require('../controllers/groupController')
+const model = require('../models/userModel');
+const schema = mongoose.model("User");
```
Comment (BuyukCem) [inline src/controllers/userController.js:3]: can you delete schema variable because you don't use this

File: src/controllers/userController.js [hunk 8]
```diff
@@ -1,145 +1,183 @@
...
+const services = require('../services/users.services')
+const bcrypt = require("bcrypt");
+const jwt = require("../utils/jwt");
+const ApplicationError = require('../errors/application.errors')
+const {userSchemaValidation, loginValidation,} = require('../utils/validationSchema.js');
+const validationParams = require('../utils/validationParams')
 
 exports.create_user = async (req, res) => {
-    console.log(req.body)
```
Comment (BuyukCem) [inline src/controllers/userController.js:11]: delete this console.log

File: src/controllers/userController.js [hunk 9]
```diff
@@ -1,145 +1,183 @@
...
-    ).catch(function(err){
-        console.log(err)
-        res.status(500)
-        res.json({message: 'Error Server'})
-    })
+    }
+};
+
+exports.logout = (req, res) => {
+    // TODO log Winston
```
Comment (BuyukCem) [inline src/controllers/userController.js:181]: please delete this comment

File: src/services/groups.services.js [hunk 10]
```diff
@@ -0,0 +1,63 @@
+const mongoose = require('mongoose');
+const model = mongoose.model("Group");
+const ifValidId = require('../utils/validationParams')
```
Comment (BuyukCem) [inline src/services/groups.services.js:3]: Delete this because you used the isValid line5 function

File: src/utils/validationParams.js [hunk 11]
```diff
@@ -0,0 +1,20 @@
...
+    return new Promise((resolve, reject) => {
+        console.log(mongoose.Types.ObjectId(id))
+        if (mongoose.Types.ObjectId(id)) {
+            resolve()
+        } else {
+            reject()
+        }
+    });
+}
+module.exports.isValid = (id) => {
```
Comment (BuyukCem) [inline src/utils/validationParams.js:18]: Please add documentation to this function

**Review by BuyukCem — CHANGES_REQUESTED** [review]
Can you correct the comments given on the lines

**Pull request merged**